### DISCUSSION TITLE
iOS cpu too high

### DISCUSSION THREAD

**Discussion by SiuyuenChow** [google/filament#6600] [General]
hello,

I have a scene that needs to render new entities every second.
These entities do not load the model directly, but are rendered through Vertex and Index.

Whenever I needed to render a new entity, I removed the old entity from the scene and added the new entity.

So the problem is, the CPU stays high.
![image](https://user-images.githubusercontent.com/12928929/221819540-f402e8ec-03b4-45a6-a6c2-3b3d3239e135.png)

I feel like it's probably a result of rendering new entities too often,
I'm trying to add entities and put them in an array, render the entities array all at once,
I see a creation method to add an array of entities, but I don't see a method to put entities in an array.

If there is a good way, please let me know.

Thanks.

code :
` struct Vertex *HDMAP_VERTICES = [self getVertexData:data];
    
    struct Indices *TRIANGLE_INDICES = [self getIndicesData:data];

    filament::VertexBuffer::BufferDescriptor vertices(HDMAP_VERTICES, sizeof(Vertex) * data.count , nullptr);
    
    filament::IndexBuffer::BufferDescriptor indices(TRIANGLE_INDICES, sizeof(uint16_t) * data.count , nullptr);

    const uint8_t stride = sizeof(Vertex);

    filament::VertexBuffer *vertexBuffer = filament::VertexBuffer::Builder()
        .vertexCount(data.count)
        .bufferCount(1)
        .attribute(filament::VertexAttribute::POSITION, 0, filament::VertexBuffer::AttributeType::FLOAT2,offsetof(Vertex, position),stride)
        .attribute(filament::VertexAttribute::COLOR, 0, filament::VertexBuffer::AttributeType::UBYTE4,offsetof(Vertex, color),stride)
        .build(*engine);

    filament::IndexBuffer *indexBuffer = filament::IndexBuffer::Builder()
        .indexCount(data.count)
        .bufferType(filament::IndexBuffer::IndexType::USHORT)
        .build(*engine);

    vertexBuffer -> setBufferAt(*engine, 0, std::move(vertices));
    indexBuffer -> setBuffer(*engine, std::move(indices));

    filamat::MaterialBuilder::init();

    filamat::Package pkg = filamat::MaterialBuilder()
        .material("void material (inout MaterialInputs material) {"
                  "prepareMaterial(material);"
                  "material.baseColor = getColor();"
                  "}")
        .depthWrite(true)
        .depthCulling(true)
        .shading(filamat::MaterialBuilder::Shading::UNLIT)
        .require(filament::VertexAttribute::COLOR)
        .targetApi(filamat::MaterialBuilder::TargetApi::METAL)
        .platform(filamat::MaterialBuilder::Platform::MOBILE)
        .build(engine->getJobSystem());

    assert(pkg.isValid());

    filamat::MaterialBuilder::shutdown();

    filament::Material *material = filament::Material::Builder().package(pkg.getData(), pkg.getSize()).build(*engine);

    filament::MaterialInstance *materialInstance =  material -> getDefaultInstance();

    utils::Entity entity;
    utils::EntityManager::get().create(1, &entity);

    filament::RenderableManager::Builder(1)
        .geometry(0, filament::RenderableManager::PrimitiveType::LINE_STRIP, vertexBuffer, indexBuffer, 0, data.count).material(0, materialInstance)
        .culling(false)
        .receiveShadows(false)
        .castShadows(false)
        .build(*engine, entity);

    engine -> getTransformManager().create(entity);

    scene -> addEntity(entity);`

**Comment by bejado** [comment]
How are you determining the CPU usage is "too high"? What's the baseline CPU usage when, for example, you're only rendering a single entity?

**Reply by SiuyuenChow** [reply]
Hello @bejado ,

I'm looking directly at xcode's cpu changes, which are consistent within a range

The following image is a single entity, the same second is removed and then added and then rendered, the cpu is only in the instant of 20% ~ 35% fluctuations, it seems OK. But if it's more (80 entities) then it goes on for another 160%, like the picture in question

![image](https://user-images.githubusercontent.com/12928929/222031169-d0ad8130-31c1-4cc7-98af-8be5a9c8e35f.png)

Thank You.

**Reply by bejado** [reply]
It makes sense that rendering more entities increases the CPU usage. Are all the entities using the same vertex/index buffers? You can try experimenting with auto-instancing:

https://github.com/google/filament/blob/78785a5bca19975852c6eb63e6a846fd453d9daf/filament/include/filament/Engine.h#L497-L511
